### DISCUSSION TITLE
How to run JBang with sudo?

### DISCUSSION THREAD

**Discussion by FDelporte** [jbangdev/jbang#1229] [Q&A] [answered]
I'm creating a Pi4J example with JBang. Problem: the underlying native libraries needs to be called with 'sudo' to control the GPIO of a Raspberry Pi, and that seems to cause problems to run JBang. Do you happen to have some kind of guidance for this specific topic? 

My current draft content: https://github.com/Pi4J/pi4j.github.io/blob/main/content-draft/jbang.md

When following this approach:
* `sudo jbang HelloWorld.java` -> jbang not found
* install jbang with sudo -> also didn't work but already forgot the error ;-) will need to check back

Sidenote: sudo is required because of how PiGpio is used as the underlying native library that actually handles the GPIO-interfacing. We are looking how we can prevent/work-around this, but are stuck with it at this moment...

**Comment by maxandersen** [comment, marked as answer]
Hmm - so the issue is that jbang is not set up in the sudo context. 

What PATH is available ?

If you installed with curl then doing “sudo ~/.jbang/bin/jbang …” should work ?

**Reply by FDelporte** [reply]
thanks will try out asap!

**Reply by maxandersen** [reply]
Reading up on how sudo behaves. It uses secure_path as opposed to the PATH. That explains why it’s missing. 

We can’t reliable install into secure_path using curl nor even most package managers so best is probably (for now) to use a full path. To simplify instructions you could use:

    sudo `which jbang`
    
As that give the full path to sudo and as such be minimal setup but also not change sudo environment. Which I think is a good thing.

**Reply by FDelporte** [reply]
this seems like a very nice idea indeed! will try out and come back here with the results

**Reply by FDelporte** [reply]
I love it when a problem can be fixed with such an elegant one-line fix! :-)
Thanks a lot, this is now an "official" part of the documentation website! 
https://pi4j.com/documentation/building/jbang/

**Reply by FDelporte** [reply]
FYI I will create a video later this week to add to this post to show all the steps to make it even more clear to new users.

**Reply by quintesse** [reply]
Awesome! :-)
